packages/cli-kit/src/public/node/themes/api.ts

```diff
@@ -77,32 +80,54 @@ export async function fetchChecksums(id: number, session: AdminSession): Promise
   return []
 }
 
-interface UpgradeThemeOptions {
-  fromTheme: number
-  toTheme: number
-  script?: string
-  session: AdminSession
-}
+export async function themeUpdate(id: number, params: ThemeParams, session: AdminSession): Promise<Theme | undefined> {
+  const name = params.name
+  if (name === undefined) {
+    throw new AbortError('Theme name is required')
+  }
 
-export async function upgradeTheme(upgradeOptions: UpgradeThemeOptions): Promise<Theme | undefined> {
-  const {fromTheme, toTheme, session, script} = upgradeOptions
-  const params = {from_theme: fromTheme, to_theme: toTheme, ...(script && {script})}
-  const response = await request('POST', `/themes`, session, params)
-  return buildTheme(response.json.theme)
-}
+  const response = await adminRequestDoc(ThemeUpdate, session, {
+    id: themeGid(id),
+    input: {name},
+  })
+  const theme = response.themeUpdate?.theme
+  if (!theme) {
+    const userErrors = response.themeUpdate?.userErrors.map((error) => error.message).join(', ')
+    throw new AbortError(userErrors as string)
+  }
 
-export async function updateTheme(id: number, params: ThemeParams, session: AdminSession): Promise<Theme | undefined> {
-  const response = await request('PUT', `/themes/${id}`, session, {theme: {id, ...params}})
-  return buildTheme(response.json.theme)
+  return buildTheme({
+    id: parseInt((theme.id as unknown as string).split('/').pop() as string, 10),
+    name: theme.name,
+    role: theme.role.toLowerCase(),
+  })
 }
 
-export async function publishTheme(id: number, session: AdminSession): Promise<Theme | undefined> {
-  return updateTheme(id, {role: 'main'}, session)
+export async function themePublish(id: number, session: AdminSession): Promise<Theme | undefined> {
+  const response = await adminRequestDoc(ThemePublish, session, {id: themeGid(id)})
+
+  const theme = response.themePublish?.theme
+  if (!theme) {
+    const userErrors = response.themePublish?.userErrors.map((error) => error.message).join(', ')
+    throw new AbortError(userErrors as string)
+  }
```

Comment:
This is somewhere between a nit and a request. Ideally we're not casting `as string`. We can rearrange this a little bit:

```suggestion
  const {themePublish} = await adminRequestDoc(ThemePublish, session, {id: themeGid(id)})

  if (themePublish.userErrors.length) {
    const userErrors = themePublish.userErrors.map((error) => error.message).join(', ')
    throw new AbortError(userErrors)
  }

  if (!themePublish.theme) {
    // This is an unknown error that should never be hit but we should support for type safety
    throw new AbortError('Failed to publish theme')
  }
```

What's really weird is that `themePublish` can be `null` or `undefined` as defined in `packages/cli-kit/src/cli/api/graphql/admin/generated/theme_publish.ts` which makes me think maybe the types are being generated wrong? I feel like the object should either exist or whatever function was trying to run should have thrown an error. Does that make sense?

Reply:
When an error is raised during execution, graphql returns a `nil` payload (separate from `user_error`), which is why `themePublish` can be `null`/`undefined`

Reply:
I like your refactor, I think there I conflated `user_error` vs execution error (return nil). I will implement similarly for the other mutations 🙏